jax/api.py

```diff
@@ -2031,3 +1502,121 @@ def fun_remat(*args, **kwargs):
+  return fun_
+
+def custom_vjp(fwd, bwd):
+  @wraps(fwd)
+  def fun_(*args, **kwargs):
+    args_flat, in_tree = tree_flatten((args, kwargs))
+    flat_fun, out_data = _flatten_fun_and_count_res(lu.wrap_init(fwd), in_tree)
+    out_flat = custom_vjp_call(flat_fun, *args_flat, bwd=bwd, out_data=out_data,
+                               keep_res=False)
+    ans_tree, _, _ = out_data()
+    return tree_unflatten(ans_tree, out_flat)
+  return fun_
+
+def _custom_call_bind(primitive, f, *args, **params):
+  top_trace = core.find_top_trace(args)
+  level = (core.trace_state.trace_stack.next_level(True)
+           if top_trace is None else top_trace.level)
+  f = _check_for_env_traces(f, level)
+  if top_trace is None:
+    with core.new_sublevel():
+      return primitive.impl(f, *args, **params)
+  else:
+    tracers = map(top_trace.full_raise, args)
+    outs = top_trace.process_custom_call(primitive, f, tracers, params)
+    outs = map(core.full_lower, outs)
+  return outs
+
+def _custom_call_impl(f, *args, **params):
+  keep_res = params.pop('keep_res')
+  bdims, stores = params.pop('bdims', ()), params.pop('stores', ())
+  out_bdims = []
+  for bd_in in bdims:
+    f, bd_out = batching.batch_fun2(f, True, bd_in)
+    out_bdims.append(bd_out)
+  outs = f.call_wrapped(*args, keep_res=keep_res)
+  for bd_out, s in zip(out_bdims, stores):
+    s.store(bd_out())
+  return outs
+
+custom_vjp_call_p = core.Primitive('custom_vjp_call')
+custom_vjp_call = partial(_custom_call_bind, custom_vjp_call_p)
+custom_vjp_call_p.def_custom_bind(custom_vjp_call)
+custom_vjp_call_p.def_impl(_custom_call_impl)
+
+custom_jvp_call_p = core.Primitive('custom_jvp_call')
+custom_jvp_call = partial(_custom_call_bind, custom_jvp_call_p)
+custom_jvp_call_p.def_custom_bind(custom_jvp_call)
+custom_jvp_call_p.def_impl(_custom_call_impl)
+
+@lu.transformation_with_aux
+def _flatten_fun_and_count_res(in_tree_, *args_flat, **params):
+  py_args, py_kwargs = tree_unflatten(in_tree_, args_flat)
+  ans, res = yield py_args, py_kwargs
+  ans_flat, ans_tree = tree_flatten(ans)
+  if params['keep_res']:
+    res = res() if callable(res) else res
+    res_flat, res_tree = tree_flatten(res)
+    num_res = len(res_flat)
+    yield res_flat + ans_flat, (ans_tree, res_tree, num_res)
+  else:
+    yield ans_flat, (ans_tree, None, 0)
+
+@lu.transformation
+def _check_for_env_traces(level, *args, **kwargs):
+  outs = yield args, kwargs
+  if any(isinstance(x, core.Tracer) and x.trace.level > level for x in outs):
+    raise Exception("closed-over traced values")
+  yield outs
+
+# This transpose rule doesn't live in ad.py because it needs batching.py
+def _custom_linearized_transpose(cts_in, *invals, **params):
+  nres, argnums, bwd, avals_out, ans_tree, res_tree, bdims_in, bdims_res, bdims_out = \
+      split_dict(params, ['num_res', 'argnums', 'bwd', 'avals_out', 'ans_tree',
+                          'res_tree', 'bdims_in', 'bdims_res', 'bdims_out'])
```

Comment:
By the way, now that we're Python 3 only you don't need to use `split_dict` anymore for keyword arguments after `*args`. You can just write something like:
```python
def _custom_linearized_transpose(
    cts_in, *invals, num_res, argnums, bwd, avals_out, ans_tree,
    res_tree, bdims_in, bdims_res, bdims_out):
```

Reply:
So cool! Thanks!